File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-3.java
class Solution {
    public int characterReplacement(String s, int k) {
        Map<Character, Integer> freq = new HashMap<>();
        int left = 0;
        int maxFreq = 0, maxLength = 1;

        for (int right = 0; right < s.length(); right++) {
            freq.put(s.charAt(right), freq.getOrDefault(s.charAt(right), 0) + 1);
            maxFreq = Math.max(maxFreq, freq.get(s.charAt(right)));

            while ((right - left + 1) - maxFreq > k) {
                freq.put(s.charAt(left), freq.get(s.charAt(left)) - 1);
                left++;
            }

            maxLength = Math.max(maxLength, right - left + 1);
        }
        
        return maxLength;
    }
}

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-3.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        
        unordered_map<char, int> freq;
        int left = 0, maxLength = 0;

        for (int right = 0; right < s.length(); right++) {
            ++freq[s[right]];

            while (freq[s[right]] > 1) {
                --freq[s[left]];
                ++left;
            }

            maxLength = max(right - left + 1, maxLength);
        }

        return maxLength;
    }
};